pkg/objstore/swift/swift.go

```diff
@@ -82,215 +190,130 @@ func (c *Container) Name() string {
+		objects, err := c.connection.ObjectNames(c.name, opts)
 		if err != nil {
-			return false, err
+			return objects, errors.Wrap(err, "swift list object names")
 		}
-		for _, objectName := range objectNames {
-			if err := f(objectName); err != nil {
-				return false, err
+		for _, object := range objects {
+			if object == SegmentsDir {
+				continue
+			}
+			if err := f(object); err != nil {
+				return objects, errors.Wrap(err, "swift iteration over objects")
 			}
 		}
-
-		return true, nil
+		return objects, nil
 	})
 }
 
-// Get returns a reader for the given object name.
-func (c *Container) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+func (c *Container) get(name string, headers swift.Headers, checkHash bool) (io.ReadCloser, error) {
 	if name == "" {
-		return nil, errors.New("error, empty container name passed")
+		return nil, errors.New("Object name cannot be empty")
 	}
-	response := objects.Download(c.client, c.name, name, nil)
-	return response.Body, response.Err
+	file, _, err := c.connection.ObjectOpen(c.name, name, checkHash, headers)
+	if err != nil {
+		return nil, errors.Wrap(err, "swift open object")
+	}
+	return file, err
 }
 
-// GetRange returns a new range reader for the given object name and range.
-func (c *Container) GetRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
-	lowerLimit := ""
-	upperLimit := ""
-	if off >= 0 {
-		lowerLimit = fmt.Sprintf("%d", off)
-	}
-	if length > 0 {
-		upperLimit = fmt.Sprintf("%d", off+length-1)
-	}
-	options := objects.DownloadOpts{
-		Newest: true,
-		Range:  fmt.Sprintf("bytes=%s-%s", lowerLimit, upperLimit),
+// Get returns a reader for the given object name.
+func (c *Container) Get(_ context.Context, name string) (io.ReadCloser, error) {
+	return c.get(name, swift.Headers{}, true)
+}
+
+func (c *Container) GetRange(_ context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	// Set Range HTTP header, see the docs https://docs.openstack.org/api-ref/object-store/?expanded=show-container-details-and-list-objects-detail,get-object-content-and-metadata-detail#id76.
+	bytesRange := fmt.Sprintf("bytes=%d-", off)
+	if length != -1 {
+		bytesRange = fmt.Sprintf("%s%d", bytesRange, off+length-1)
 	}
-	response := objects.Download(c.client, c.name, name, options)
-	return response.Body, response.Err
+	return c.get(name, swift.Headers{"Range": bytesRange}, false)
 }
 
 // Attributes returns information about the specified object.
-func (c *Container) Attributes(ctx context.Context, name string) (objstore.ObjectAttributes, error) {
-	response := objects.Get(c.client, c.name, name, nil)
-	headers, err := response.Extract()
+func (c *Container) Attributes(_ context.Context, name string) (objstore.ObjectAttributes, error) {
+	if name == "" {
+		return objstore.ObjectAttributes{}, errors.New("Object name cannot be empty")
```

Comment:
```suggestion
		return objstore.ObjectAttributes{}, errors.New("object name cannot be empty")
```

Reply:
fixed in https://github.com/thanos-io/thanos/pull/2732/commits/7a51f585a7184cc9a0c5a73ffcd77b23896da026